File: bot-api/dotnet/test/src/test_utils/MockedServerTest.cs
using System.Threading.Tasks;
using NUnit.Framework;
using Robocode.TankRoyale.BotApi.Tests.Test_utils;

namespace Robocode.TankRoyale.BotApi.Tests
{
    public class MockedServerTest
    {
        private MockedServer _server;

        [SetUp]
        public void SetUp()
        {
            _server = new MockedServer();
            _server.Start();
        }

        [TearDown]
        public void TearDown()
        {
            _server.Stop();
        }

        [Test]
        public void AwaitBotReady_ShouldSucceed()
        {
            var bot = new TestBot();
            Task.Run(bot.Start);

            bool ready = _server.AwaitBotReady(30000);
            Assert.That(ready, Is.True, "Bot should be ready");
        }

        [Test]
        public void SetBotStateAndAwaitTick_ShouldUpdateStateAndReturnTrue()
        {
            var bot = new TestBot();
            Task.Run(bot.Start);

            Assert.That(_server.AwaitBotReady(30000), Is.True, "Bot should be ready");

            // Start the bot's game loop
            Task.Run(bot.Go);

            double newEnergy = 50.0;
            double newSpeed = 4.0;

            bool success = _server.SetBotStateAndAwaitTick(newEnergy, null, newSpeed, null, null, null);

            Assert.That(success, Is.True, "SetBotStateAndAwaitTick should succeed");

            // Wait a bit for the bot to process the tick
            System.Threading.Thread.Sleep(200);

            Assert.That(bot.Energy, Is.EqualTo(newEnergy));
            Assert.That(bot.Speed, Is.EqualTo(newSpeed));
        }

        private class TestBot : BaseBot
        {
            public TestBot() : base(BotInfo.Builder()
                .SetName("TestBot")
                .SetVersion("1.0")
                .AddAuthor("Author")
                .SetDescription("Description")
                .SetHomepage("https://test.com")
                .AddCountryCode("us")
                .AddGameType("classic")
                .SetPlatform(".NET")
                .SetProgrammingLang("C#")
                .Build(), MockedServer.ServerUrl)
            {
            }
        }
    }
}

File: bot-api/java/src/test/java/test_utils/MockedServerTest.java
package test_utils;

import dev.robocode.tankroyale.botapi.BaseBot;
import dev.robocode.tankroyale.botapi.BotInfo;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class MockedServerTest {

    private MockedServer server;

    @BeforeEach
    void setUp() {
        server = new MockedServer();
        server.start();
    }

    @AfterEach
    void tearDown() {
        server.stop();
    }

    @Test
    void awaitBotReady_shouldSucceed() {
        var bot = new TestBot();
        new Thread(bot::start).start();

        boolean ready = server.awaitBotReady(30000);
        assertThat(ready).as("Bot should be ready").isTrue();
    }

    @Test
    void setBotStateAndAwaitTick_shouldUpdateStateAndReturnTrue() {
        var bot = new TestBot();
        new Thread(bot::start).start();

        assertThat(server.awaitBotReady(30000)).as("Bot should be ready").isTrue();

        double newEnergy = 50.0;
        double newSpeed = 4.0;

        boolean success = server.setBotStateAndAwaitTick(newEnergy, null, newSpeed, null, null, null);

        assertThat(success).as("setBotStateAndAwaitTick should succeed").isTrue();
        assertThat(bot.getEnergy()).isEqualTo(newEnergy);
        assertThat(bot.getSpeed()).isEqualTo(newSpeed);
    }

    private static class TestBot extends BaseBot {
        TestBot() {
            super(BotInfo.builder()
                    .setName("TestBot")
                    .setVersion("1.0")
                    .addAuthor("Author")
                    .setDescription("Description")
                    .setHomepage("https://test.com")
                    .addCountryCode("us")
                    .addGameType("classic")
                    .setPlatform("JVM")
                    .setProgrammingLang("Java")
                    .build(), MockedServer.getServerUrl());
        }
    }
}
